// @amityco/ts-sdk/src/channel/observers/liveChannels.ts
/* eslint-disable no-use-before-define */
import { getResolver } from '~/core/model';
import { pullFromCache, pushToCache } from '~/cache/api';
import { getActiveClient } from '~/client/api';
import {
  createQuery,
  runQuery,
  queryOptions,
  filterByPropEquality,
  filterByChannelMembership,
  sortByLastCreated,
  sortByDisplayName,
  sortByFirstCreated,
} from '~/core/query';
import { onChannelCreated, onChannelDeleted, onChannelUpdated } from '../events';
import { queryChannels } from '../api';

import {
  COLLECTION_DEFAULT_CACHING_POLICY,
  COLLECTION_DEFAULT_PAGINATION_LIMIT,
  ENABLE_CACHE_MESSAGE,
} from '~/utils/constants';

/**
 * ```js
 * import { liveChannels } from '@amityco/ts-sdk'
 *
 * let channels = []
 * const unsub = liveChannels({
 *   displayName: Amity.Channel['displayName'],
 * }, response => merge(channels, response.data))
 * ```
 *
 * Observe all mutations on a list of {@link Amity.Channel}s
 *
 * @param params for querying channels
 * @param callback the function to call when new data are available
 * @returns An {@link Amity.Unsubscriber} function to run when willing to stop observing the channels
 *
 * @category Channel Live Collection
 */
export const liveChannels = (
  params: Amity.ChannelLiveCollection,
  callback: Amity.LiveCollectionCallback<Amity.Channel>,
  config?: Amity.LiveCollectionConfig,
) => {
  const { log, cache, userId } = getActiveClient();

  if (!cache) {
    console.log(ENABLE_CACHE_MESSAGE);
  }

  const timestamp = Date.now();
  log(`liveChannels(tmpid: ${timestamp}) > listen`);

  const { limit: queryLimit, ...queryParams } = params;

  const limit = queryLimit ?? COLLECTION_DEFAULT_PAGINATION_LIMIT;
  const { policy = COLLECTION_DEFAULT_CACHING_POLICY } = config ?? {};

  const disposers: Amity.Unsubscriber[] = [];
  const cacheKey = ['channel', 'collection', {}];

  const responder = (data: Amity.ChannelLiveCollectionCache) => {
    let channels: Amity.Channel[] =
      data.data
        .map(channelId => pullFromCache<Amity.Channel>(['channel', 'get', channelId])!)
        .filter(Boolean)
        .map(({ data }) => data) ?? [];

    channels = filterByPropEquality(channels, 'isDeleted', params.isDeleted);
    channels = filterByPropEquality(channels, 'displayName', params.displayName);

    if (params.types) {
      channels = channels.filter(c => params.types?.includes(c.type));
    }

    if (params.tags) {
      channels = channels.filter(c => c.tags?.some(t => params.tags?.includes(t)));
    }

    if (params.excludeTags) {
      channels = channels.filter(c => !c.tags?.some(t => params.excludeTags?.includes(t)));
    }

    // userId is required to be able to filter channel by current active user
    if (params.membership && userId) {
      channels = filterByChannelMembership(channels, params.membership, userId);
    }

    const sortBy = params.sortBy || 'lastCreated';

    if (sortBy === 'lastCreated' || sortBy === 'firstCreated') {
      channels = channels.sort(sortBy === 'lastCreated' ? sortByLastCreated : sortByFirstCreated);
    }

    /*
     * The server returns channels with empty | null displayName's first before
     * returning sorted list of channels with displayNames
     *
     * This section needs to be updated as displayNames can be null as well
     */
    if (sortBy === 'displayName') {
      channels = channels
        // this needs to be aligned with the backend data type
        .map(c => (c.displayName ? c : { ...c, displayName: '' }))
        // @ts-ignore
        .sort(sortByDisplayName);
    }

    callback({
      onNextPage: onFetch,
      data: channels,
      hasNextPage: !!data.params?.page,
      loading: data.loading,
      error: data.error,
    });
  };

  const realtimeRouter = (channel: Amity.Channel, action: Amity.ChannelActionType) => {
    const collection = pullFromCache<Amity.ChannelLiveCollectionCache>(cacheKey)?.data;
    if (!collection) return;

    if (action === 'onCreate') {
      collection.data = [...new Set([channel.channelId, ...collection.data])];
      /*
       * if the filter isDeleted is set, only collection item must be updated and not
       * the collection
       */
    } else if (action === 'onDelete' && !params.isDeleted) {
      collection.data = collection.data.filter(c => c !== channel.channelId);
    }

    pushToCache(cacheKey, collection);
    responder(collection);
  };

  const onFetch = () => {
    const collection = pullFromCache<Amity.ChannelLiveCollectionCache>(cacheKey)?.data;

    const channels = collection?.data ?? [];

    if (channels.length > 0 && !collection?.params.page) return;

    const query = createQuery(queryChannels, {
      ...queryParams,
      page: collection?.params.page ?? { limit },
    });

    runQuery(
      query,
      ({ data: result, error, loading, nextPage: page }) => {
        const data = {
          loading,
          error,
          params: { page },
          data: channels,
        };

        if (result) {
          data.data = [...new Set([...channels, ...result.map(getResolver('channel'))])];
        }

        pushToCache(cacheKey, data);

        responder(data);
      },
      queryOptions(policy),
    );
  };

  disposers.push(
    onChannelCreated(channel => realtimeRouter(channel, 'onCreate')),
    onChannelDeleted(channel => realtimeRouter(channel, 'onDelete')),
    onChannelUpdated(channel => realtimeRouter(channel, 'onUpdate')),
  );

  onFetch();

  return () => {
    log(`liveChannels(tmpid: ${timestamp}) > dispose`);
    disposers.forEach(fn => fn());
  };
};
