// @amityco/ts-sdk/src/utils/tests/dummy/channel.ts
import { date, rolesAndPermissions, user11, user12, user13 } from '.';

const mockPage = {
  paging: {
    previous: 'eyJiZWZvcmUiOjU1LCJsYXN0IjoxMH0=',
    next: 'eyJiZWZvcmUiOjU1LCJsYXN0IjoxMH0=',
  },
};

export const channelRaw1: Amity.ChannelRaw = {
  tags: ['tag2'],
  messageCount: 1,
  createdAt: date,
  isDeleted: false,
  type: 'community',
  channelId: 'channelId11',
};

export const channelRaw2: Amity.ChannelRaw = {
  tags: ['tag1'],
  messageCount: 3,
  createdAt: date,
  isDeleted: true,
  type: 'conversation',
  channelId: 'channelId12',
};

export const channelRaw3: Amity.ChannelRaw = {
  messageCount: 3,
  createdAt: date,
  type: 'community',
  channelId: 'channelId21',
};

export const channel11: Amity.Channel = {
  ...channelRaw1,
  hasMention: false,
  unreadCount: 0,
  type: 'community',
};

export const channel12: Amity.Channel = {
  ...channelRaw2,
  hasMention: true,
  unreadCount: 2,
};

export const channel21: Amity.Channel = {
  ...channelRaw3,
  hasMention: false,
  unreadCount: 0,
  type: 'conversation',
  isDeleted: true,
  tags: ['tag1'],
};

export const channelUser1: Amity.Membership<'channel'> = {
  isMuted: false,
  createdAt: date,
  muteTimeout: '',
  isBanned: false,
  readToSegment: 1,
  lastActivity: date,
  membership: 'member',
  userId: user11.userId,
  ...rolesAndPermissions,
  lastMentionedSegment: 0,
  channelId: channel11.channelId,
};

export const channelUser2: Amity.Membership<'channel'> = {
  ...channelUser1,
  readToSegment: 1,
  lastMentionedSegment: 2,
  channelId: channel12.channelId,
};

export const channelUser3: Amity.Membership<'channel'> = {
  isMuted: false,
  createdAt: date,
  muteTimeout: '',
  isBanned: false,
  readToSegment: 1,
  lastActivity: date,
  userId: user12.userId,
  ...rolesAndPermissions,
  membership: 'none',
  roles: ['test-role'],
  lastMentionedSegment: 0,
  channelId: channel11.channelId,
};

export const channelUser4: Amity.Membership<'channel'> = {
  ...channelUser3,
  userId: user13.userId,
};

export const channelUser5: Amity.Membership<'channel'> = {
  isMuted: false,
  createdAt: date,
  muteTimeout: '',
  isBanned: true,
  readToSegment: 1,
  lastActivity: date,
  userId: user13.userId,
  ...rolesAndPermissions,
  membership: 'banned',
  lastMentionedSegment: 0,
  channelId: channel11.channelId,
};

export const channelUser6: Amity.Membership<'channel'> = {
  isMuted: true,
  createdAt: date,
  muteTimeout: '',
  isBanned: true,
  readToSegment: 1,
  lastActivity: date,
  userId: 'test',
  ...rolesAndPermissions,
  membership: 'banned',
  lastMentionedSegment: 0,
  channelId: channel11.channelId,
};

export const channelUser7: Amity.Membership<'channel'> = {
  ...channelUser3,
  membership: 'member',
};

export const channelQueryResponse = {
  data: {
    channels: [channel11, channel12],
    channelUsers: [],
    files: [],
    users: [],
    ...mockPage,
  },
};

export const channelQueryResponsePage2 = {
  data: {
    channels: [channel21],
    channelUsers: [],
    files: [],
    users: [],
    ...mockPage,
  },
};

export const channelUserQueryResponse = {
  data: {
    channels: [channel11],
    channelUsers: [channelUser1, channelUser3],
    users: [],
    files: [],
    ...mockPage,
  },
};

export const channelUserQueryResponsePage2 = {
  data: {
    channels: [channel11],
    channelUsers: [channelUser4],
    users: [],
    files: [],
    ...mockPage,
  },
};

export const channelUserQueryResponsePage3 = {
  data: {
    channels: [channel11],
    channelUsers: [channelUser7, channelUser6],
    users: [],
    files: [],
    ...mockPage,
  },
};
