// @amityco/ts-sdk/src/utils/tests/dummy/post.ts
import { encodeJson } from '..';
import { user11, date } from '.';

export const post11: Amity.Post = {
  path: '',
  feedId: '',
  children: [],
  comments: [],
  flagCount: 0,
  reactions: {},
  hashFlag: null,
  editedAt: date,
  createdAt: date,
  commentsCount: 0,
  reactionsCount: 0,
  postId: 'postId11',
  targetType: 'user',
  parentId: 'postId12',
  data: { text: 'post11' },
  targetId: user11.userId,
  hasFlaggedComment: false,
  hasFlaggedChildren: false,
  postedUserId: user11.userId,
};

export const post12: Amity.Post = {
  ...post11,
  postId: 'post12',
};

export const post13: Amity.Post = {
  ...post11,
  postId: 'post13',
};

export const posts = {
  targetId: user11.userId,
  targetType: 'user' as Amity.PostTargetType,
  page1: [post11.postId, post12.postId, post13.postId],
  page2: ['postId21', 'postId22', 'postId23'],
  page3: ['postId31', 'postId32', 'postId33'],
};

const nextPage = encodeJson({ before: post13.postId, limit: 10 });

export const postQueryResponse = {
  data: {
    posts: [post11, post12, post13],
    files: [],
    paging: {
      next: nextPage,
    },
  },
};
