// @amityco/ts-sdk/src/core/subscription.ts
import { getActiveClient } from '~/client/api/activeClient';
import { ASCError } from '~/core/errors';
import { proxyMqttEvents } from '~/core/events';
import { getActiveUser } from '~/client/api/activeUser';

export enum SubscriptionLevels {
  COMMUNITY = 'community',
  POST = 'post',
  COMMENT = 'comment',
  POST_AND_COMMENT = 'post_and_comment',
  USER = 'user',
}

const getCommunityUserTopic = (
  path: Amity.Subscribable['path'],
  level?: SubscriptionLevels,
): string => {
  switch (level) {
    case 'post':
      return `${path}/post/+`;
    case 'comment':
      return `${path}/post/+/comment/+`;
    case 'post_and_comment':
      return `${path}/post/#`;
    default:
      return path;
  }
};

export const getCommunityTopic = (
  { path }: Amity.Subscribable,
  level: Exclude<SubscriptionLevels, SubscriptionLevels.USER> = SubscriptionLevels.COMMUNITY,
): string => getCommunityUserTopic(path, level);

export const getUserTopic = (
  { path }: Amity.Subscribable,
  level: Exclude<SubscriptionLevels, SubscriptionLevels.COMMUNITY> = SubscriptionLevels.USER,
): string =>
  getCommunityUserTopic(
    level === SubscriptionLevels.USER ? path : path.replace(/^(\w*)/, '$1/social'),
    level,
  );

export const getPostTopic = (
  { path }: Amity.Subscribable,
  level: SubscriptionLevels.POST | SubscriptionLevels.COMMENT = SubscriptionLevels.POST,
): string => {
  switch (level) {
    case 'comment':
      return `${path}/comment/+`;
    default:
      return path;
  }
};

export const getCommentTopic = ({ path }: Amity.Subscribable): string => {
  return path;
};

export const getMyFollowersTopic = (): string => {
  const user = getActiveUser();

  const [networkId] = user.path.split('/user/');

  return `${networkId}/membership/${user._id}/+/+`;
};

export const getMyFollowingsTopic = (): string => {
  const user = getActiveUser();

  const [networkId] = user.path.split('/user/');

  return `${networkId}/membership/+/${user._id}/+`;
};

/** @hidden */
export const getNetworkTopic = (): string => {
  const user = getActiveUser();

  const [networkId] = user.path.split('/user/');

  return networkId;
};

let mqttAccessToken: string;
let mqttUserId: string;

export function subscribeTopic(
  topic: string,
  callback?: Amity.Listener<ASCError | void>,
): Amity.Unsubscriber {
  const { mqtt, emitter, accessToken } = getActiveClient();
  const user = getActiveUser();

  if (mqttAccessToken !== accessToken || mqttUserId !== user._id) {
    mqttAccessToken = accessToken!;
    mqttUserId = user._id!;

    mqtt.connect({ accessToken: mqttAccessToken, userId: mqttUserId });
    proxyMqttEvents(mqtt, emitter);
  }

  return mqtt.subscribe(topic, callback);
}
